src/test/java/lambda/part3/exercise/Exercise2.java

```diff
@@ -0,0 +1,129 @@
+package lambda.part3.exercise;
+
+import java.util.stream.Collectors;
+import lambda.data.Employee;
+import lambda.data.JobHistoryEntry;
+import lambda.data.Person;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.junit.Assert.assertEquals;
+
+@SuppressWarnings({"unused", "WeakerAccess", "ConstantConditions"})
+public class Exercise2 {
+
+    private static class MapHelper<T> {
+
+        private final List<T> source;
+
+        private MapHelper(List<T> source) {
+            this.source = source;
+        }
+
+        public static <T> MapHelper<T> from(List<T> source) {
+            return new MapHelper<>(source);
+        }
+
+        public List<T> getMapped() {
+            return new ArrayList<>(source);
+        }
+
+        public <R> MapHelper<R> map(Function<T, R> mapping) {
+            List<R> resultList = source.stream()
+                .map(mapping)
+                .collect(Collectors.toList());
+            return new MapHelper<R>(resultList);
+        }
+
+        public <R> MapHelper<R> flatMap(Function<T, List<R>> flatMapping) {
+            List<R> resultList = source.stream()
+                .map(flatMapping)
+                .flatMap(l->l.stream())
+                .collect(Collectors.toList());
+            return new MapHelper<R>(resultList);
+        }
+    }
+
+    @Test
+    public void mapEmployeesToLengthOfTheirFullNamesUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> lengths = MapHelper.from(employees)
+                .map(Employee::getPerson)
+                .map(Person::getFullName)
+                .map(String::length)
+                .getMapped();
+        assertEquals(Arrays.asList(14, 19, 14, 15, 14, 16), lengths);
+    }
+
+    @Test
+    public void mapEmployeesToCodesOfLetterTheirPositionsUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> codes = MapHelper.from(employees)
+        .flatMap(Employee::getJobHistory)
+        .map(JobHistoryEntry::getPosition)
+        .flatMap(Exercise2::calcCodes)
+        .getMapped();
```

Comment:
As I know you should write here
```
// TODO  .map(JobHistoryEntry -> String(position))
// TODO  .flatMap(String -> Character(letter))
// TODO  .map(Character -> Integer(code letter)
// TODO  .getMapped();
```
But you used `Exercise2::calcCodes` instead of  your functions.

Reply:
Yep, you are right. Fixed